Sources/AsyncQueue/MainActorQueue.swift

```diff
@@ -0,0 +1,110 @@
+// MIT License
+//
+// Copyright (c) 2023 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/// A queue that enables enqueing ordered asynchronous tasks from a nonisolated context onto the `@MainActor`'s isolated context.
+/// Tasks are guaranteed to begin executing in the order in which they are enqueued. However, if a task suspends it will allow subsequently enqueued tasks to begin executing.
+/// This queue exhibits the execution behavior of an actor: tasks sent to this queue can re-enter the queue, and tasks may execute in non-FIFO order when a task suspends.
+///
+/// A `MainActorQueue` ensures tasks sent from a nonisolated context to the `@MainActor`'s isolated context begin execution in order.
+public final class MainActorQueue: Sendable {
```

Comment:
We're not at the point where we could just have `MainActorQueue` be a special case of the regular actor queue could we?

Reply:
It *could* be a special case of `ActorQueue`, but the biggest difference is that the `MainActorQueue` runs its `for await task in taskStream` _on `Main`_ rather than on some background queue. Limiting queue hops makes us slightly more efficient / more akin to `DispatchQueue.main.async` in terms of performance.

We also mark every closure as being a `@MainActor` closure, which helps ensure the compiler does the right thing / always keeps us on `Main`. (I recently found an issue where lack of explicitness cause some code to run off of `Main` that should have been running on `Main`, so I default now to "be explicit if possible")

Reply:
> Limiting queue hops makes us slightly more efficient / more akin to `DispatchQueue.main.async` in terms of performance.

I just ran some tests and `MainActorQueue` still manages to have a queue hop from the `Task { @MainActor in` call – the `Task` initially starts running on a background queue and then enters `Main` before executing our code (per the debugger), so we're still slower than `DispatchQueue.main.async`. Maaaaybe I take a pass at making this implemented with an `ActorQueue` 🤔

Reply:
I also just tried to implement `MainActorQueue` utilizing an `ActorQueue` and remembered why this doesn't work:
`ActorQueue` adopts the execution context of an `Actor`-conforming type, and `final class MainActorQueue` does not conform to `Actor`!

Heck, even `@MainActor final class MainActorQueue` yields a `Non-actor type 'MainActorQueue' cannot conform to the 'Actor' protocol` error.

Reply:
It's ok, this might be doable a few Swift revisions from now but it's clear we're going to be hitting some pain points.

Reply:
So, actually doable:
```
diff --git a/Sources/AsyncQueue/MainActorQueue.swift b/Sources/AsyncQueue/MainActorQueue.swift
index 3dde0b7..5a4adf0 100644
--- a/Sources/AsyncQueue/MainActorQueue.swift
+++ b/Sources/AsyncQueue/MainActorQueue.swift
@@ -31,23 +31,7 @@ public final class MainActorQueue: Sendable {
 
     /// Instantiates a main actor queue.
     init() {
-        var capturedTaskStreamContinuation: AsyncStream<@Sendable @MainActor () async -> Void>.Continuation? = nil
-        let taskStream = AsyncStream<@Sendable @MainActor () async -> Void> { continuation in
-            capturedTaskStreamContinuation = continuation
-        }
-        // Continuation will be captured during stream creation, so it is safe to force unwrap here.
-        // If this force-unwrap fails, something is fundamentally broken in the Swift runtime.
-        taskStreamContinuation = capturedTaskStreamContinuation!
-
-        Task.detached { @MainActor in
-            for await task in taskStream {
-                await MainActor.shared.suspendUntilStarted(task)
-            }
-        }
-    }
-
-    deinit {
-        taskStreamContinuation.finish()
+        actorQueue.adoptExecutionContext(of: MainActor.shared)
     }
 
     // MARK: Public
@@ -59,7 +43,9 @@ public final class MainActorQueue: Sendable {
     /// The scheduled task will not execute until all prior tasks have completed or suspended.
     /// - Parameter task: The task to enqueue.
     public func enqueue(_ task: @escaping @Sendable @MainActor () async -> Void) {
-        taskStreamContinuation.yield(task)
+        actorQueue.enqueue { _ in
+            await task()
+        }
     }
 
     /// Schedules an asynchronous task and returns after the task is complete.
@@ -67,10 +53,8 @@ public final class MainActorQueue: Sendable {
     /// - Parameter task: The task to enqueue.
     /// - Returns: The value returned from the enqueued task.
     public func enqueueAndWait<T: Sendable>(_ task: @escaping @Sendable @MainActor () async -> T) async -> T {
-        await withUnsafeContinuation { continuation in
-            taskStreamContinuation.yield {
-                continuation.resume(returning: await task())
-            }
+        await actorQueue.enqueueAndWait { _ in
+            await task()
         }
     }
 
@@ -79,33 +63,12 @@ public final class MainActorQueue: Sendable {
     /// - Parameter task: The task to enqueue.
     /// - Returns: The value returned from the enqueued task.
     public func enqueueAndWait<T: Sendable>(_ task: @escaping @Sendable @MainActor () async throws -> T) async throws -> T {
-        try await withUnsafeThrowingContinuation { continuation in
-            taskStreamContinuation.yield {
-                do {
-                    continuation.resume(returning: try await task())
-                } catch {
-                    continuation.resume(throwing: error)
-                }
-            }
+        try await actorQueue.enqueueAndWait { _ in
+            try await task()
         }
     }
 
     // MARK: Private
 
-    private let taskStreamContinuation: AsyncStream<@Sendable @MainActor () async -> Void>.Continuation
-}
-
-extension MainActor {
-    @MainActor
-    fileprivate func suspendUntilStarted(_ task: @escaping @Sendable @MainActor () async -> Void) async {
-        // Suspend the calling code until our enqueued task starts.
-        await withUnsafeContinuation { continuation in
-            // Utilize the serial (but not FIFO) @MainActor context to execute the task without requiring the calling method to wait for the task to complete.
-            Task { @MainActor in
-                // Signal that the task has started. Since this `task` is executing on the main actor's execution context, the order of execution is guaranteed.
-                continuation.resume()
-                await task()
-            }
-        }
-    }
+    private let actorQueue = ActorQueue<MainActor>()
 }
```
But def more queue hops – will keep this as-is for now.